File: Tests/TokenSpendieTests/ClaudeProviderTests.swift
import XCTest
@testable import TokenSpendie

final class ClaudeProviderTests: XCTestCase {

    // --- Test doubles ---
    final class StubCredentials: CredentialStore {
        var result: Result<OAuthCredentials, Error>
        var exists: Bool
        var loadCount = 0
        init(_ result: Result<OAuthCredentials, Error>, exists: Bool = true) {
            self.result = result
            self.exists = exists
        }
        func loadCredentials() throws -> OAuthCredentials {
            loadCount += 1
            return try result.get()
        }
        func credentialsExist() -> Bool { exists }
    }
    final class StubEndpoint: ClaudeUsageEndpoint {
        var results: [Result<UsageSnapshot, Error>]
        var callCount = 0
        init(_ results: [Result<UsageSnapshot, Error>]) { self.results = results }
        func fetchUsage(accessToken: String) async throws -> UsageSnapshot {
            defer { callCount += 1 }
            return try results[min(callCount, results.count - 1)].get()
        }
    }

    private func creds() -> OAuthCredentials {
        OAuthCredentials(accessToken: "tok", refreshToken: nil, expiresAt: nil)
    }
    private func usageSnapshot() -> UsageSnapshot {
        UsageSnapshot(
            session: UsageWindow(percent: 47, resetsAt: Date(timeIntervalSince1970: 100)),
            weekly: UsageWindow(percent: 31, resetsAt: Date(timeIntervalSince1970: 200)),
            modelWeeklies: [ModelWeekly(model: "Opus",
                                        window: UsageWindow(percent: 62, resetsAt: nil))],
            fetchedAt: Date(timeIntervalSince1970: 999))
    }

    // --- detectCredentials ---

    func testDetectCredentialsReflectsKeychainExistence() {
        let present = ClaudeProvider(credentials: StubCredentials(.success(creds()), exists: true),
                                     endpoint: StubEndpoint([]))
        XCTAssertTrue(present.detectCredentials())
        let absent = ClaudeProvider(credentials: StubCredentials(.success(creds()), exists: false),
                                    endpoint: StubEndpoint([]))
        XCTAssertFalse(absent.detectCredentials())
    }

    // --- conversion ---

    func testConvertMapsWindowsWithLabelsAndHeadline() {
        let snapshot = ClaudeProvider.convert(usageSnapshot())
        XCTAssertEqual(snapshot.id, .claude)
        XCTAssertEqual(snapshot.headline.label, "Session · 5h")
        XCTAssertEqual(snapshot.headline.window.percent, 47, accuracy: 0.001)
        XCTAssertEqual(snapshot.windows.map(\.label),
                       ["Session · 5h", "Weekly · all", "Weekly · Opus"])
        XCTAssertEqual(snapshot.windows[0].resetStyle, .countdown)
        XCTAssertEqual(snapshot.windows[1].resetStyle, .date)
        XCTAssertEqual(snapshot.windows[1].detail, "all models")
        XCTAssertEqual(snapshot.windows[2].detail, "Opus only")
        XCTAssertEqual(snapshot.fetchedAt, Date(timeIntervalSince1970: 999))
        XCTAssertNil(snapshot.plan)
    }

    // --- fetchUsage ---

    func testFetchUsageReturnsConvertedSnapshot() async throws {
        let provider = ClaudeProvider(credentials: StubCredentials(.success(creds())),
                                      endpoint: StubEndpoint([.success(usageSnapshot())]))
        let snapshot = try await provider.fetchUsage()
        XCTAssertEqual(snapshot.headline.window.percent, 47, accuracy: 0.001)
    }

    func testFetchUsageRetriesOnceByRereadingKeychainOn401() async throws {
        let credentials = StubCredentials(.success(creds()))
        let endpoint = StubEndpoint([.failure(ProviderError.unauthorized),
                                     .success(usageSnapshot())])
        let provider = ClaudeProvider(credentials: credentials, endpoint: endpoint)
        _ = try await provider.fetchUsage()
        XCTAssertEqual(credentials.loadCount, 2, "re-reads the Keychain once on 401")
        XCTAssertEqual(endpoint.callCount, 2)
    }

    func testFetchUsagePropagatesPersistentUnauthorized() async {
        let provider = ClaudeProvider(
            credentials: StubCredentials(.success(creds())),
            endpoint: StubEndpoint([.failure(ProviderError.unauthorized),
                                    .failure(ProviderError.unauthorized)]))
        do {
            _ = try await provider.fetchUsage()
            XCTFail("expected unauthorized")
        } catch {
            XCTAssertEqual(error as? ProviderError, .unauthorized)
        }
    }

    func testFetchUsagePropagatesCredentialError() async {
        let provider = ClaudeProvider(
            credentials: StubCredentials(.failure(CredentialError.notFound)),
            endpoint: StubEndpoint([]))
        do {
            _ = try await provider.fetchUsage()
            XCTFail("expected notFound")
        } catch {
            XCTAssertEqual(error as? CredentialError, .notFound)
        }
    }
}

File: windows/tests/TokenSpendie.Windows.Tests/Data/ClaudeProviderTests.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using NSubstitute;
using NSubstitute.ExceptionExtensions;
using TokenSpendie.Windows.Data;
using TokenSpendie.Windows.Models;
using Xunit;

namespace TokenSpendie.Windows.Tests.Data;

public class ClaudeProviderTests
{
    private static OAuthCredentials Creds(string token = "tok") =>
        new(token, null, null);

    private static UsageSnapshot Snapshot() => new(
        Session: new UsageWindow(47, DateTimeOffset.FromUnixTimeSeconds(100)),
        Weekly: new UsageWindow(31, DateTimeOffset.FromUnixTimeSeconds(200)),
        ModelWeeklies: new[] { new ModelWeekly("Opus", new UsageWindow(62, null)) },
        FetchedAt: DateTimeOffset.FromUnixTimeSeconds(999));

    [Fact]
    public void IdAndDisplayName()
    {
        var provider = new ClaudeProvider(
            Substitute.For<ICredentialReader>(),
            Substitute.For<IClaudeUsageEndpoint>());
        provider.Id.Should().Be(ProviderID.Claude);
        provider.DisplayName.Should().Be("Claude");
    }

    [Fact]
    public void DetectCredentialsDelegates()
    {
        var reader = Substitute.For<ICredentialReader>();
        reader.CredentialsExist().Returns(true);
        new ClaudeProvider(reader, Substitute.For<IClaudeUsageEndpoint>()).DetectCredentials()
            .Should().BeTrue();
        reader.CredentialsExist().Returns(false);
        new ClaudeProvider(reader, Substitute.For<IClaudeUsageEndpoint>()).DetectCredentials()
            .Should().BeFalse();
    }

    [Fact]
    public void ConvertMapsWindowsWithLabelsAndHeadline()
    {
        var snapshot = ClaudeProvider.Convert(Snapshot());
        snapshot.Id.Should().Be(ProviderID.Claude);
        snapshot.Headline.Label.Should().Be("Session");
        snapshot.Headline.Window.Percent.Should().BeApproximately(47, 0.001);
        snapshot.Windows.Select(w => w.Label).Should().Equal("Session", "Weekly", "Weekly · Opus");
        snapshot.Windows[0].ResetStyle.Should().Be(ResetStyle.Countdown);
        snapshot.Windows[1].ResetStyle.Should().Be(ResetStyle.Date);
        snapshot.Windows[1].Detail.Should().Be("all models");
        snapshot.Windows[2].Detail.Should().Be("Opus only");
        snapshot.FetchedAt.Should().Be(DateTimeOffset.FromUnixTimeSeconds(999));
        snapshot.Plan.Should().BeNull();
    }

    [Fact]
    public async Task FetchUsageReturnsConvertedSnapshot()
    {
        var reader = Substitute.For<ICredentialReader>();
        reader.LoadCredentialsAsync(Arg.Any<CancellationToken>()).Returns(Creds());
        var endpoint = Substitute.For<IClaudeUsageEndpoint>();
        endpoint.FetchUsageAsync(Arg.Any<string>(), Arg.Any<CancellationToken>())
            .Returns(Snapshot());

        var provider = new ClaudeProvider(reader, endpoint);
        var snapshot = await provider.FetchUsageAsync();

        snapshot.Headline.Window.Percent.Should().BeApproximately(47, 0.001);
    }

    [Fact]
    public async Task FetchUsageRetriesOnceByRereadingCredentialsOn401()
    {
        var reader = Substitute.For<ICredentialReader>();
        reader.LoadCredentialsAsync(Arg.Any<CancellationToken>()).Returns(Creds());

        var endpoint = Substitute.For<IClaudeUsageEndpoint>();
        endpoint.FetchUsageAsync(Arg.Any<string>(), Arg.Any<CancellationToken>())
            .Returns(
                _ => throw new ProviderUnauthorizedException(),
                _ => Task.FromResult(Snapshot()));

        var provider = new ClaudeProvider(reader, endpoint);
        _ = await provider.FetchUsageAsync();

        await reader.Received(2).LoadCredentialsAsync(Arg.Any<CancellationToken>());
        await endpoint.Received(2).FetchUsageAsync(Arg.Any<string>(), Arg.Any<CancellationToken>());
    }

    [Fact]
    public async Task FetchUsagePropagatesPersistentUnauthorized()
    {
        var reader = Substitute.For<ICredentialReader>();
        reader.LoadCredentialsAsync(Arg.Any<CancellationToken>()).Returns(Creds());
        var endpoint = Substitute.For<IClaudeUsageEndpoint>();
        endpoint.FetchUsageAsync(Arg.Any<string>(), Arg.Any<CancellationToken>())
            .Throws(new ProviderUnauthorizedException());

        var provider = new ClaudeProvider(reader, endpoint);
        Func<Task> act = () => provider.FetchUsageAsync();
        await act.Should().ThrowAsync<ProviderUnauthorizedException>();
    }

    [Fact]
    public async Task FetchUsagePropagatesCredentialError()
    {
        var reader = Substitute.For<ICredentialReader>();
        reader.LoadCredentialsAsync(Arg.Any<CancellationToken>())
            .Throws(new CredentialNotFoundException());
        var endpoint = Substitute.For<IClaudeUsageEndpoint>();

        var provider = new ClaudeProvider(reader, endpoint);
        Func<Task> act = () => provider.FetchUsageAsync();
        await act.Should().ThrowAsync<CredentialNotFoundException>();
        await endpoint.DidNotReceive().FetchUsageAsync(Arg.Any<string>(), Arg.Any<CancellationToken>());
    }
}
